### PR TITLE
Feat: v3 Java

### PR DESCRIPTION
**Pull request by Meldiron**
Java implementation of v3. Parent PR: https://github.com/open-runtimes/open-runtimes/pull/136

### CODE REVIEW

File: runtimes/java-18.0/src/main/java/io/openruntimes/java/Server.java [hunk 1]
```diff
@@ -9,53 +9,158 @@
...
         try {
-            RuntimeResponse userResponse = codeWrapper.main(request, response);
-            Map<String, Object> output = new HashMap<String, Object>();
-            output.put("response", userResponse.data);
-            output.put("stdout", outStream.toString());
-            return resp.result(RuntimeResponse.gson.toJson(output));
+            Class classToLoad = Class.forName("io.openruntimes.java.Tests");
+            Method classMethod = classToLoad.getDeclaredMethod("main", RuntimeContext.class);
+            Object instance = classToLoad.newInstance();
+            output = (RuntimeOutput) classMethod.invoke(instance, context);
```
Comment (abnegate) [inline runtimes/java-18.0/src/main/java/io/openruntimes/java/Server.java:125]: Does this need to be updated from the `Tests` class? You shouldn't need to use reflection here

File: runtimes/java-18.0/src/main/java/io/openruntimes/java/Server.java [hunk 2]
```diff
@@ -9,53 +9,158 @@
...
+            output = context.res.send("", 500, new HashMap<String, String>());
         } finally {
             System.out.flush();
             System.err.flush();
             System.setOut(systemOut);
             System.setErr(systemErr);
         }
+
+        if(output == null) {
+            context.error("Return statement missing. return context.res.empty() if no response is expected.");
```
Comment (abnegate) [inline runtimes/java-18.0/src/main/java/io/openruntimes/java/Server.java:141]: If `output` is null here the function must have returned `null`; the message will be misleading

File: runtimes/java-18.0/build.sh [hunk 3]
```diff
@@ -16,26 +16,6 @@ for filename in ./*.gradle*; do
...
-# Wrap the user code in a class
-echo "package io.openruntimes.java;
-${IMPORTS}
-public class Wrapper {
-${CODE}
-}" > Wrapper.java
-
-# Remove the user code file (copy)
-rm "${INTERNAL_RUNTIME_ENTRYPOINT}"
-
```
[the remark this answers is not in the document]
Comment (Meldiron) [reply runtimes/java-18.0/build.sh:38]: User is required to write the whole class in his code.

File: runtimes/java-18.0/src/main/java/io/openruntimes/java/RuntimeContext.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+            RuntimeRequest req,
+            RuntimeResponse res
+    ) {
+        this.req = req;
+        this.res = res;
+    }
+
+    public void log(Object message) {
+        if(message instanceof Map || message instanceof List) {
+            Gson gson = new GsonBuilder().serializeNulls().create();
```
[the remark this answers is not in the document]
Comment (Meldiron) [reply runtimes/java-18.0/src/main/java/io/openruntimes/java/RuntimeContext.java:29]: Updated in all places

File: runtimes/java-18.0/src/main/java/io/openruntimes/java/RuntimeResponse.java [hunk 5]
```diff
@@ -1,44 +1,47 @@
...
+    public RuntimeOutput send(String body, int statusCode) {
+        return this.send(body, statusCode, new HashMap<String, String>());
+    }
+    public RuntimeOutput send(String body) {
+        return this.send(body, 200, new HashMap<String, String>());
+    }
 
-    static final Gson gson = new Gson();
+    public RuntimeOutput json(Map<String, Object> json, int statusCode, Map<String, String> headers) {
+        Gson gson = new GsonBuilder().serializeNulls().create();
```
[the remark this answers is not in the document]
Comment (Meldiron) [reply runtimes/java-18.0/src/main/java/io/openruntimes/java/RuntimeResponse.java:21]: Updated in all places

File: runtimes/java-18.0/src/main/java/io/openruntimes/java/Server.java (hunk 2 above)
Comment (Meldiron) [reply runtimes/java-18.0/src/main/java/io/openruntimes/java/Server.java:141]: I am making myself note to change the message in all runtimes. Let's ignore it as part of this PR

File: runtimes/java-18.0/src/main/java/io/openruntimes/java/Server.java [hunk 6]
```diff
@@ -9,53 +9,158 @@
...
 
     public static Resp execute(Req req, Resp resp) {
-        if (!req.headers().containsKey("x-internal-challenge") || req.headers().get("x-internal-challenge").isEmpty()) {
-            return resp.code(500).result("Unauthorized");
+        Map<String, String> reqHeaders = req.headers();
+
+        String serverSecret = System.getenv("OPEN_RUNTIMES_SECRET");
+        if(serverSecret == null) {
+            serverSecret = "";
+        }
```
[the remark this answers is not in the document]
Comment (Meldiron) [reply runtimes/java-18.0/src/main/java/io/openruntimes/java/Server.java:39]: Are you sure? What if both variable is empty string and header is empty string? In such a scenario we need to make sure to return 500.

File: runtimes/java-18.0/src/main/java/io/openruntimes/java/RuntimeContext.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+import com.google.gson.GsonBuilder;
+import com.google.gson.Gson;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class RuntimeContext {
+    public RuntimeRequest req;
+    public RuntimeResponse res;
```
[the remark this answers is not in the document]
Comment (Meldiron) [reply runtimes/java-18.0/src/main/java/io/openruntimes/java/RuntimeContext.java:14]: If we do that, `context.req` becomes `context.getRe()`. I am not sure if we like it.

File: runtimes/java-18.0/src/main/java/io/openruntimes/java/RuntimeRequest.java [hunk 8]
```diff
@@ -5,60 +5,42 @@
...
+    public Object body;
+    public Map<String, String> headers;
+    public String method;
+    public String url;
+    public String scheme;
+    public String host;
+    public String path;
+    public int port;
+    public String queryString;
+    public Map<String, String> query;
```
[the remark this answers is not in the document]
Comment (Meldiron) [reply runtimes/java-18.0/src/main/java/io/openruntimes/java/RuntimeRequest.java:18]: If we do that, `context.req.method` becomes `context.req.getMethod()`. I am not sure if we like it.

File: runtimes/java-18.0/src/main/java/io/openruntimes/java/RuntimeOutput.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
+package io.openruntimes.java;
+
+import java.util.Map;
+
+public class RuntimeOutput {
+    public String body;
+    public int statusCode;
+    public Map<String, String> headers;
```
[the remark this answers is not in the document]
Comment (Meldiron) [reply runtimes/java-18.0/src/main/java/io/openruntimes/java/RuntimeOutput.java:8]: Applied

File: runtimes/java-18.0/src/main/java/io/openruntimes/java/Server.java (hunk 1 above)
Comment (Meldiron) [reply runtimes/java-18.0/src/main/java/io/openruntimes/java/Server.java:125]: Good catch, I hardcoded this and forgot about it. Updated.

File: runtimes/java-18.0/src/main/java/io/openruntimes/java/RuntimeContext.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (abnegate) [reply runtimes/java-18.0/src/main/java/io/openruntimes/java/RuntimeContext.java:14]: It's a choice between consistency and familiarity. Personally I would expect `getReq()` for Java, directly accessing a public field would feel out of place

File: runtimes/java-18.0/src/main/java/io/openruntimes/java/RuntimeRequest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (abnegate) [reply runtimes/java-18.0/src/main/java/io/openruntimes/java/RuntimeRequest.java:18]: I will open a discussion to get more feedback

File: runtimes/java-18.0/src/main/java/io/openruntimes/java/Server.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (abnegate) [reply runtimes/java-18.0/src/main/java/io/openruntimes/java/Server.java:40]: Shouldn't that be okay? Do we disallow setting the secret to an empty string?

[the remark this answers is not in the document]
Comment (Meldiron) [reply runtimes/java-18.0/src/main/java/io/openruntimes/java/Server.java:40]: We decided to require secret env variable for security reasons. Open Runtimes should be only used for server-to-server communication where secrets are present. In future, we can introduce load balancer to take care of secrets generation and management to allow public executions.

File: runtimes/java-18.0/src/main/java/io/openruntimes/java/RuntimeContext.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (Meldiron) [reply runtimes/java-18.0/src/main/java/io/openruntimes/java/RuntimeContext.java:14]: Applied.

File: runtimes/java-18.0/src/main/java/io/openruntimes/java/RuntimeRequest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (Meldiron) [reply runtimes/java-18.0/src/main/java/io/openruntimes/java/RuntimeRequest.java:18]: Applied.

**Pull request merged**